modules/top-level/test.nix

```diff
@@ -8,8 +8,72 @@
 let
   cfg = config.test;
 
-  inherit (config) warnings;
-  assertions = builtins.concatMap (x: lib.optional (!x.assertion) x.message) config.assertions;
+  # Expectation submodule used for checking warnings and/or assertions
+  expectation = lib.types.submodule (
+    { config, ... }:
+    {
+      options = {
+        predicate = lib.mkOption {
+          type = with lib.types; functionTo bool;
+          description = ''
+            A predicate of that determines whether this expectation is met.
+
+            Type
+            ```
+            [String] -> Boolean
+            ```
+
+            Parameters
+            1. values - all warnings or matched assertions
+          '';
+        };
+        expect = lib.mkOption {
+          type = with lib.types; nullOr (enum (builtins.attrNames cfg.predefinedExpectationPredicates));
+          description = ''
+            If non-null, will be used together with `value` to define `predicate`.
+          '';
+          default = "any";
+        };
+        value = lib.mkOption {
+          type = lib.types.anything;
+          description = ''
+            If defined, will be used together with `assertion` to define `assertionFunction`.
+          '';
+        };
+        message = lib.mkOption {
+          type = with lib.types; either str (functionTo str);
+          description = ''
+            The assertion message.
+
+            If the value is a function, it is called with the same list of warning/assertion messages that is applied to `predicate`.
+          '';
+          defaultText = lib.literalMD ''
+            If `assertion` is non-null, a default is computed. Otherwise, there is no default.
+          '';
+        };
+      };
+      config = lib.mkIf (config.expect != null) {
+        predicate = lib.mkOptionDefault (
+          cfg.predefinedExpectationPredicates.${config.expect}.predicate config.value
+        );
+        message = lib.mkOptionDefault (
+          cfg.predefinedExpectationPredicates.${config.expect}.message config.value
+        );
+      };
+    }
+  );
+
+  expectationType =
+    with lib.types;
+    coercedTo (either str ints.unsigned) (
+      value:
+      {
+        inherit value;
+      }
+      // lib.optionalAttrs (builtins.isInt value) {
+        expect = "count";
```

Comment:
For the same reason of the default I'm not too keen on having this coerction. I'd prefer we always specify what kind of assertion we want. If needs be we can define helper functions to reduce the boilerplate, like `expectCount = n: { expect = "count"; value = n; }`, or even `mkExpect = expect: value: { inherit expect value; }`;

Reply:
I had a similar thought process but went in a different direction. I'd rather avoid introducing such helper functions as I feel `coercedTo` is somewhat self-documenting.

Although I agree type-dependent coercion could be kinda confusing. I think this can still be documented well enough? Not convinced either way.

Reply:
I quite dislike automatic coercions in most cases, I always prefer explicit ways, as I find myself always forgetting about coercions I rarely use